src/link/NvPtx.zig

```diff
@@ -0,0 +1,115 @@
+const Air = @import("../Air.zig");
+const Liveness = @import("../Liveness.zig");
+const LlvmObject = @import("../codegen/llvm.zig").Object;
+
+base: link.File,
+llvm_object: *LlvmObject,
+
+pub fn createEmpty(gpa: *Allocator, options: link.Options) !*NvPtx {
+    const nvptx = try gpa.create(NvPtx);
+    nvptx.* = .{
+        .base = .{
+            .tag = .nvptx,
+            .options = options,
+            .file = null,
+            .allocator = gpa,
+        },
+        .llvm_object = undefined,
+    };
+
+    if (!build_options.have_llvm) return error.TODOArchNotSupported;
+
+    switch (options.target.cpu.arch) {
+        .nvptx, .nvptx64 => {},
+        else => return error.TODOArchNotSupported,
+    }
+
+    switch (options.target.os.tag) {
+        // TODO: does it also work with nvcl ?
+        .cuda => {},
+        else => return error.TODOOsNotSupported,
+    }
+
+    return nvptx;
+}
+
+pub fn openPath(allocator: *Allocator, sub_path: []const u8, options: link.Options) !*NvPtx {
+    assert(options.object_format == .nvptx);
+    if (!build_options.have_llvm or !options.use_llvm) return error.NvptxRequiresLlvm;
+
+    const nvptx = try createEmpty(allocator, options);
+    errdefer nvptx.base.destroy();
+    log.info("Opening .ptx target file {s}", .{sub_path});
+    nvptx.llvm_object = try LlvmObject.create(allocator, sub_path, options);
+    return nvptx;
+}
+
+pub fn deinit(self: *NvPtx) void {
+    self.llvm_object.destroy(self.base.allocator);
+}
+
+pub fn updateFunc(self: *NvPtx, module: *Module, func: *Module.Fn, air: Air, liveness: Liveness) !void {
+    try self.llvm_object.updateFunc(module, func, air, liveness);
+}
+
+pub fn updateDecl(self: *NvPtx, module: *Module, decl: *Module.Decl) !void {
+    return self.llvm_object.updateDecl(module, decl);
+}
+
+pub fn updateDeclExports(
+    self: *NvPtx,
+    module: *Module,
+    decl: *const Module.Decl,
+    exports: []const *Module.Export,
+) !void {
+    if (build_options.skip_non_native and builtin.object_format != .nvptx) {
+        @panic("Attempted to compile for object format that was disabled by build configuration");
+    }
+    return self.llvm_object.updateDeclExports(module, decl, exports);
+}
+
+pub fn freeDecl(self: *NvPtx, decl: *Module.Decl) void {
+    return self.llvm_object.freeDecl(decl);
+}
+
+pub fn flush(self: *NvPtx, comp: *Compilation) !void {
+    return self.flushModule(comp);
+}
+
+pub fn flushModule(self: *NvPtx, comp: *Compilation) !void {
+    if (build_options.skip_non_native or !build_options.have_llvm) {
+        @panic("Attempted to compile for architecture that was disabled by build configuration");
+    }
+    const tracy = trace(@src());
+
+    var hack_comp = comp;
+    if (comp.bin_file.options.emit) |emit| {
+        hack_comp.emit_asm = .{
+            .directory = emit.directory,
+            .basename = self.llvm_object.sub_path,
+        };
+        hack_comp.bin_file.options.emit = null;
+    }
```

Comment:
Another option is to just override the values for `emit_bin` and `emit_asm` set in main.zig, though that might get just as messy. What does clang even do when you do not pass `-femit-asm`?

Reply:
so without `-femit-asm` Clang will create a .o object that contains:

* the .ptx embedded in a binary format (.fatbin)
* some functions to init the Cuda module that mostly delegate to `cudart` (Cuda Runtime library) 
* stubs for the device functions

But note this is a Clang thing, LLVM doesn't know about cudart, and will fail if you pass it `emit-bin`: "LLVM failed to emit asm=(none) bin=kernel.ptx.ptx ir=(none) bc=(none): TargetMachine can't emit an object file"